### ISSUE TITLE
Position hexes correctly when border width is not 0

### ISSUE THREAD

**Issue by grandsong** [flauwekeul/honeycomb#16]
After successfuly following the examples, I got a pointy grid of hexes with stroke width = 1.
But soon I noticed that the left strokes of the first column is thinner than others.

I changed stroke width into much larger values and realized that **half** of those strokes are out of the left side of the viewport of the svg root/container element.
The same story happens to the top corner strokes which stand partly beyond the top side of the viewport.

So, when calculating positions of hexes, stroke/border width ought to be used.

---

I've fixed this issue.

The screenshot shows the dashed outline (set via css) of the container svg element.
The borders & corners of hexes are right in the box of the element.

![grid-of-hexes-with-border](https://user-images.githubusercontent.com/3063419/46201021-b4e70980-c345-11e8-8426-c99861cddea2.jpg)

Here's the code:

```js

  Point = Honeycomb.Point;
  sqrt3 = Math.sqrt(3);

  Hex = Honeycomb.extendHex({
    border: 12,
    .
    .
    .
    toPoint: function() {
      var border, q, r, size, x, y;
      {q, r, size, border} = this;
      border = border || 0;
      if (this.isPointy()) {
        x = size * sqrt3 * (q + r / 2) + border / 2;
        y = size * 3 / 2 * r + border / sqrt3;
      } else {
        x = size * 3 / 2 * q + border / sqrt3;
        y = size * sqrt3 * (r + q / 2) + border / 2;
      }
      return Point(x, y);
    }

```

I have also worked a lot to make the **width and height of the viewport** be exactly of the visual size of the grid, so that the right and bottom borders/corners can sit next to the sides of the viewport too.

Here's the essential parts of my work. Hope you can understand them without much pain...

```js
  l = {
    grid: {
      w: 4,
      h: 3
    }
  };
  .
  .
  .
  el = document.querySelector '#main-grid'
  draw = SVG el


  grid = Grid.rectangle({
    width: l.grid.w,
    height: l.grid.h,
    onCreate: function(hex) {
      return hex.render(draw);
    }
  });


  hex_proto = Hex().__proto__;

  if (hex_proto.isPointy()) {
    grid.view = {
      w: hex_proto.size * sqrt3 * (l.grid.w + 0.5) + hex_proto.border,
      h: hex_proto.size * (l.grid.h / 2 * 3 + 0.5) + hex_proto.border * 2 / sqrt3
    };
  } else {
    grid.view = {
      h: hex_proto.size * sqrt3 * (l.grid.h + 0.5) + hex_proto.border,
      w: hex_proto.size * (l.grid.w / 2 * 3 + 0.5) + hex_proto.border * 2 / sqrt3
    };
  }
  el.style.width = `${grid.view.w}px`;
  el.style.height = `${grid.view.h}px`;

```

**Comment by flauwekeul** [comment]
Thanks for being so thorough in explaining your work.

I'm not sure what you're asking of me though 😅. Do you propose to include a `border` property in `Hex` and have some methods (like `Hex#toPoint()`) take this into account?

**Comment by grandsong** [comment]
Yes.
You can call it strokeWidth, etc, though.

The issue is:
In your example, your `render` method has `.stroke({ width: 1, color: '#999' })`, but in your `toPoint`, you seem to assume that all hexes have zero width of strokes.
Set a larger (say, 12) width and you will see it obviously.

**Comment by flauwekeul** [comment]
Honeycomb is a viewless library. Rendering hexes and grids is up to the user. Adding a `border`/`strokeWidth` property to `Hex` is a view detail; it's out of scope of Honeycomb.

For example: what if you need 2 or 3 borders. Does `toPoint` need to take that into account too? What about some margin between hexes?

I think there's a solution to your problem without the need to add new properties to hexes or even without having to write your own `toPoint` method: just offset each hex with your border width. See this example: https://jsfiddle.net/Flauwekeul/8h0facvo/

**Comment by grandsong** [comment]
#### 1. But I think this lib is not viewless.

`toPoint` convert hex coordination into a pixel based one, which is used in view.
And `Grid.pointToHex()` does the rather oppsite.

When most, if not all, users use this lib, they do make views.
This is especially true since we are using Javascript here, mostly because we are to show hexes in browsers.

"border" (or, edge-width? Sorry but I'm not a native English speaker) is a very common need in real practices. And your example is, very, an example of this need.

#### 2. Yes, there are other cases.

Margins are indeed common, too.
2 or 3 borders? Too rare and too specific.

Anyway, it will be wise to provide a more generic way to cover all this sort of things.

Still, border and margin can cover 98% of normal needs, and they may be used in the same time in some cases, so I suggest providing the two, and only the two.

#### 3. It is unfriendly to leave the job to users.

If a user has to do the job himself, he will have to repeat my little (or, not so little) struggle with where to use `sqrt3` and ` / 2`.

Moreover, pointy and flat layouts are different.

The lib save people from math chores, which most people can do themselves but it would be a great waste to repeat and would be prone to errors.
This is a bless.
Then, why exclude borders/margins?

#### Conclusion

I believe most users of this lib will expect, with good reasons, that it offers an ease-to-use solution dealing with borders/margins/whatever-are-commonly-needed-and-messes-the-positions-of-hexes.

**Comment by flauwekeul** [comment]
Shall we continue this discussion in the broader issue #21? Whether to add properties like `border` or `margin` to hex instances is a matter of model-view separation and you've opened another issue to discuss that already.

**Comment by grandsong** [comment]
Even if model-view separation is done, HexView and GridView still need to deal with them...
...as well as pixel size (width and height) or whole grids, and the origins of the grids.
I've just tried `Grid::hexagon` and had to do a lot of calcuations and coding...
But let's postpone this issue.

**Comment by flauwekeul** [comment]
I totally agree a user shouldn't have to do their own calculations involving `Math.sqrt()` and the like. The example I posted earlier (https://jsfiddle.net/Flauwekeul/8h0facvo/) doesn't require these calculations and should achieve your goal of having borders not be cut off. Could you show a scenario where you feel the need to do your own calculations?

**Comment by grandsong** [comment]
1. In `render`, every user has to repeat your work, which should be simplied into one or two method calls.
2. You didn't use `sqrt3` only because you overlooked the missing part of the topmost corners, which exceed the top side of the grid a little. (If it is not obvious to you, try increasing `BORDER_WIDTH` to `40` or more.)
3. So, the offset of x and y are not the same (you thought they are both `BORDER_WIDTH / 2`). As a result, pointy and flat orientations requires different calculations.
4. `BORDER_WIDTH` still needs to be stored somewhere. I believe It should be a sibling of `size` of hex.
5. The bottom and right sides of the grid are also needed to calculate...
6. rectangle, triangle, hexagon grids have very different algorithms...
7. And it is very natual to want the origin (point (0, 0)) of a hexagon grid to be in the heart to the grid visually.

Some users can be lazy and just set a large enough left and top margins (and width and height) of the grid, but others like me want to get precise pixels.

For example, I need my grid be flexible, whose hex numbers (the `radius` of hexagon grids will increase with difficulties of level in my puzzle game) can vary, but the grid is always show in the center of the main part of the screen, which is also flexible.
(The <svg> is an inline element, so I set the style of its parent "text-align: center").

**Comment by flauwekeul** [comment]
1. There are simply too many ways to render hex grids. People might want to render actual hexagons (like in my example), or maybe circles, or any other shape. In other words: how each hex/tile is rendered is completely up to the user. I've tried to only add functions that most users need: getting corners of each hex, getting the coordinates of each hex. Users should be able to render any kind of hex with those functions. But maybe I'm misunderstanding you. Could you show what those 1 or 2 methods would do that you're referring to?
1. I see what you mean. Those top corners are cut-off because there the border is wider than on the sides. This is partly due to how strokes are rendered in svgs: their origin is always the border of the shape. There's no way to have the strokes on the [outside or inside](https://stackoverflow.com/questions/7241393/can-you-control-how-an-svgs-stroke-width-is-drawn) of the shape. The point I'm trying to make is that it's a 100% specific to SVGs. If we were using canvas to render the grid, we'd have more (at least different) control on how borders are rendered. To me, this is another example why things like borders are to remain in the user's code, not in honeycomb.
1. As pointed out above, you're right that offsetting `x` and `y` with `BORDER_WIDTH / 2` isn't correct (for `y` at least). Some more math is needed. Or a different solution altogether, e.g.: setting the CSS property `overflow: visible` on the svg element. But, as you said, that might be a "lazy" solution.
1. with `extendHex()` it's possible to set any custom properties:
    ```js
    extendHex({ size: 5, borderWidth: 10 })
    ```
1. I could add methods to get a grid's width and height, let me know if you need those. For reasons I've explained above, it will be up to the user to add any borders/margins.
1. I think the problem with the cut off borders is due to other reasons than how the hex positions are calculated (see point 2). Therefore, it doesn't matter the other shape methods have different algorithms.
1. Why would it be "very natural" the `0, 0` hex is in the middle of the grid? I think it's at least as natural to have `0, 0` hex in the top left corner.

In conclusion: I still strongly prefer to keep rendering outside of honeycomb. I hope you understand why, though you might not agree with me.

**Comment by grandsong** [comment]
1. My recent thought is: there are two grids, logical and visual. Our confusion comes from the failure to discriminate them. I'll explain this big topic in a new post.

2. Calculating the precise size on screen of a grid, when borders (and margins) matter, is really a challenge, which I was not expecting at all.
Before that moment, I had been very happy with such a convenient lib and been believing it would save me from any such annoying calculation due to the shape of hexagon and the organization of the shapes in "pointy/flat" orientations.
I was not suggesting you to integrate such calculation directly in the current grid class, which is mostly viewless. I was suggesting you provide it as a side product, shipped with the lib, so that users like me don't have to repeat it.
Again, now I think there should be two grids and therefore this issue will be reloved with the visual one.

5. Let's start with visual size of a grid without borders. The "pointy/flat" orientations are already not easy for lib users.

7. Well, I meant that in cube, not semi-rectangle, the origin is usually at center.

**Comment by flauwekeul** [comment]
1. Is this a different topic than #21 ?
2. I can very well imagine you expected honeycomb to offer more view-related functionality than it does. New projects could be made, each with its own rendering "engine": one for SVG, one for react maybe and one for canvas. Or maybe more generic: one for component-based rendering (suitable for SVG or react) and one for canvas-based rendering. I don't think it's wise to make a single project that handles all kinds of rendering, there's too much difference between component- and canvas-rendering.

I feel we've exhausted the topic about whether to add `border` and/or `margin` properties to Hex. Shall we close this issue?

**Comment by grandsong** [comment]
1. I'm writing a very long new topic. Be warned. 
But you don't need to read and answer quickly for I don't want to give you burdens.
2. I don't think you need to go that far. You only need to provide calculations relating hexagons and then users can draw hexes based on the data. The data should be generic. It is lib users' duty to convert them into specific visual objects, whether in SVG or in canvas.
Yes, you can close this issue.

**Comment by flauwekeul** [comment]
1. Ok 🙂Please don't be too elaborate. I appreciate I can take my time to reply, but I have to reply some day and my spare time is very limited. So it's in both our best interests to keep our communications brief... Then again, if you need to be elaborate to get your point across, so be it.
1. I thought honeycomb already provided that functionality 😅
1. I see your ideas in the other topic 👍
